Binary stl starting with solid

nlharr opened the pull request:
A binary STL with the first 5 bytes spelling the word "solid" can cause the STL parser to get into an infinite loop.  It would look for the word "facet", not find it, and then continue looking.  It would keep doing this even at the end of the file.  It would read nothing, see that nothing was not "facet", and then continue reading nothing.

This change limits the number of times it can search for the word "facet".  If that fails, it assumes it's a binary STL and attempts to parse that.

Commit:
Made the stl decoder swap back to assuming it is a binary stl if it cannot find the "facet" word

Commit:
added a unit test

jcz-carbon3d commented on src/draco/io/stl_decoder.cc:
```diff
@@ -62,21 +62,32 @@ Status StlDecoder::ParseHeader(bool* is_binary) {
   if (! buffer_.Decode(ascii_buffer, 5)) {
     return Status(Status::IO_ERROR, "STL file has invalid header.");
   }
+  bool is_ascii_file = ! strncmp(ascii_buffer, "solid ", 5);
```
I think I have seen ascii stl files in the wild that start with "solid\n" or "solid\r".

jcz-carbon3d commented on src/draco/io/stl_decoder.cc:
```diff
@@ -62,21 +62,32 @@ Status StlDecoder::ParseHeader(bool* is_binary) {
   if (! buffer_.Decode(ascii_buffer, 5)) {
     return Status(Status::IO_ERROR, "STL file has invalid header.");
   }
+  bool is_ascii_file = ! strncmp(ascii_buffer, "solid ", 5);
```
I think I have seen ascii stl files in the wild that start with "solid\n" or "solid\r".

nlharr replied:
The trailing space is a typo and should not affect the result of the function since solid has 5 characters and this is using strncmp, but good catch.

jcz-carbon3d commented on src/draco/io/stl_decoder.cc:
```diff
@@ -62,21 +62,32 @@ Status StlDecoder::ParseHeader(bool* is_binary) {
-  } else {
+      // If the file is not composed of a series of strings or we loop too much on this, it is likely
+      // a binary stl.
+      if (buffer_seek_point == buffer_.decoded_size() || loop_counter > 4) {
```
I am having some trouble reading this loop. Does "buffer_seek_point == buffer_.decoded_size()" mean EOF or something else?

nlharr replied:
It means either EOF, or that the parser::ParseString call has not advanced the buffer point which happens if the remaining characters are whitespace characters.

jcz-carbon3d commented on src/draco/io/stl_decoder_test.cc:
```diff
@@ -72,12 +72,15 @@ class StlDecoderTest : public ::testing::Test {
 TEST_F(StlDecoderTest, ValidBinarySTL) {
   test_decoding("teapot.stl", 946);
   test_decoding("40mmcube.stl", 12);
+  test_decoding("square_15_15_250.STL", 12);
```
Can you add a test with binary stl file that starts with the 80 bytes:

"solid q\nfacet normal 0 0 0\nouter loop\nvertex 0 0 0\nvertex 1 1 1\nvertex 1 1 0\n\n\n\n"

? This seems a bit silly, but so does a binary stl file that starts with "solid square_15_15_250                                                          ".

nlharr replied:
Done.  Fair warning, I'm kind of sick.  While attempting to spell "through" I wrote "throw" so give it a bit of a close read.

Commit:
- Fixed a merge error where the stl tests were not running.
- Made the stl decoder work with a binary stl file that's first 80 bytes are a valid ascii stl file describing facets.

Commit:
Moved the ascii parsing to it's own function

Commit:
Style

Commit:
removed is  ascii

Commit:
lint

Commit:
lint and comments

jcz-carbon3d commented on src/draco/io/stl_decoder.cc:
```diff
@@ -163,53 +174,85 @@ Status StlDecoder::ParseAsciiFace(Vector3f* v0, Vector3f* v1, Vector3f* v2, Vect
+    bool attempt_binary = false;
+    // Fills tmp_three_vec_storage with the face data.
+    status = ParseAsAscii(&attempt_binary, &tmp_three_vec_storage, &num_faces);
+    // if the ascii parsing fales in a known way fall back to parsing it as a binary stl file
```
fails

jcz-carbon3d commented on src/draco/io/stl_decoder.cc:
```diff
@@ -163,53 +174,85 @@ Status StlDecoder::ParseAsciiFace(Vector3f* v0, Vector3f* v1, Vector3f* v2, Vect
+    do {
+      status = ParseAsciiFace(&tmp_v0, &tmp_v1, &tmp_v2, &tmp_norm, &is_valid_triangle);
+      if (! status.ok()) {
+        *should_attempt_binary = is_first_solid;
```
You never set this to false in this function. You do in the current calling function so it works for now but seems risky and might break easily after future code changes.

Commit:
Nits and a fix for a bug
